### PR TITLE
lobby search

### LINKED ISSUE #147: [Feature Request]: Lobby search
### What should be implemented?

Filter the lobby list according to the content of the search text box

### PR DESCRIPTION
**Pull request by JayPi4c**
## What type of PR is this? (keep all applicable)

- Feature

## Description

Lobbies may now be searched using the `search lobbies` TextField. The textfield has been enhanced with auto-completion functionality which allows to quickly search for all existing lobbies. Searching for no lobby name will show all lobbies.

## Related Tickets & Documents

- Closes #147

### CODE REVIEW

**Review by Til7701 — CHANGES_REQUESTED** [review]
I'll take a closer look at this at a later time. I like the autocomplete feature, but when I type something in, I have to press enter three times to actually search for what I typed in. That's a bit too much. Also, I think you have to dispose of the old autocomplete object. That might be a memory leak.

**Comment by JayPi4c** [comment]
In my testing, only hitting enter twice was needed in order to perform a search. That is one for accepting the auto completion and one for really performing the search. But I agree, that it feels kinda pesky having to hit enter twice to get to the result. I'll try to change it to automatically perform a search whenever there has been a change in the text field.

About the potential memory leak: this is a really good find. I did some dive into the Java docs and I guess, it would be better to manually dispose the binding. Although I'm not completely sure that maybe also the GC will get rid of it, as it has no references to the running application. I think, you are right and we should take care of it. After all there is a special dispose method already implemented and we can easily use that.

File: client/src/main/java/org/schlunzis/kurtama/client/fx/controller/MainMenuController.java [hunk 1]
```diff
@@ -56,6 +64,30 @@ protected void updateItem(LobbyInfo lobbyInfo, boolean empty) {
                 joinLobby();
         });
         joinLobbyButton.setDisable(true);
+
+        // whenever there is a change in the lobby list, update the auto-completion
+        AtomicReference<AutoCompletionBinding<String>> bindingRef = new AtomicReference<>();
```
Comment (JayPi4c) [inline client/src/main/java/org/schlunzis/kurtama/client/fx/controller/MainMenuController.java]: Using this `AtomicReference` was the first somewhat reasonable approach I could find, to tackle the problem with effectively final variables in lambdas. If you know something better, or even better, a best practice, please let me know.

**Comment by JayPi4c** [comment]
I also did some basic testing without the call of the `dispose()` method and I could not register an increase in memory usage, so I guess, the garbage collector is doing it's job. But anyway, now we can be sure, that the old binding is removed. Who knows. Maybe I just did not notice the small footprint of the bindings.

**Review by Til7701 — APPROVED** [review]

**Pull request merged**